flac.go

```diff
@@ -64,79 +74,122 @@ func New(r io.Reader) (stream *Stream, err error) {
+	block, err := stream.parseStreamInfo()
+	if err != nil {
+		return stream, err
+	}
+
+	for !block.IsLast {
+		block, err = meta.Parse(stream.r)
+		if err != nil {
+			if err != meta.ErrReservedType {
+				return stream, err
+			} else {
+				if err = block.Skip(); err != nil {
+					return stream, err
+				}
+			}
+		}
+
+		if block.Header.Type == meta.TypeSeekTable {
+			stream.seekTable = block.Body.(*meta.SeekTable)
+		}
+	}
+
+	stream.dataStart, err = rs.Seek(0, io.SeekCurrent)
+	return stream, err
+}
+
+var (
+	// flacSignature marks the beginning of a FLAC stream.
+	flacSignature = []byte("fLaC")
 
-// id3Signature marks the beginning of an ID3 stream, used to skip over ID3 data.
-var id3Signature = []byte("ID3")
+	// id3Signature marks the beginning of an ID3 stream, used to skip over ID3 data.
+	id3Signature = []byte("ID3")
+
+	ErrInvalidSeek   = errors.New("stream.Seek: out of stream seek")
+	ErrNoSeeker      = errors.New("stream.Seek: no a Seeker")
+	ErrInvalidOption = errors.New("stream.New: invalid Option")
+)
 
 // parseStreamInfo verifies the signature which marks the beginning of a FLAC
 // stream, and parses the StreamInfo metadata block. It returns a boolean value
 // which specifies if the StreamInfo block was the last metadata block of the
 // FLAC stream.
-func (stream *Stream) parseStreamInfo() (isLast bool, err error) {
+func (stream *Stream) parseStreamInfo() (block *meta.Block, err error) {
 	// Verify FLAC signature.
 	r := stream.r
 	var buf [4]byte
 	if _, err = io.ReadFull(r, buf[:]); err != nil {
-		return false, err
+		return block, err
 	}
 
 	// Skip prepended ID3v2 data.
 	if bytes.Equal(buf[:3], id3Signature) {
 		if err := stream.skipID3v2(); err != nil {
-			return false, err
+			return block, err
 		}
 
 		// Second attempt at verifying signature.
 		if _, err = io.ReadFull(r, buf[:]); err != nil {
-			return false, err
+			return block, err
 		}
 	}
 
 	if !bytes.Equal(buf[:], flacSignature) {
-		return false, fmt.Errorf("flac.parseStreamInfo: invalid FLAC signature; expected %q, got %q", flacSignature, buf)
+		return block, fmt.Errorf("flac.parseStreamInfo: invalid FLAC signature; expected %q, got %q", flacSignature, buf)
 	}
 
 	// Parse StreamInfo metadata block.
-	block, err := meta.Parse(r)
+	block, err = meta.Parse(r)
 	if err != nil {
-		return false, err
+		return block, err
 	}
 	si, ok := block.Body.(*meta.StreamInfo)
 	if !ok {
-		return false, fmt.Errorf("flac.parseStreamInfo: incorrect type of first metadata block; expected *meta.StreamInfo, got %T", si)
+		return block, fmt.Errorf("flac.parseStreamInfo: incorrect type of first metadata block; expected *meta.StreamInfo, got %T", si)
 	}
 	stream.Info = si
-	return block.IsLast, nil
+	return block, nil
 }
 
 // skipID3v2 skips ID3v2 data prepended to flac files.
 func (stream *Stream) skipID3v2() error {
-	r := bufio.NewReader(stream.r)
-
 	// Discard unnecessary data from the ID3v2 header.
-	if _, err := r.Discard(2); err != nil {
```

Comment:
These changes were the only way I could get all the tests to pass.  I don't really understand why though :-(

Reply:
I reverted it back, and it works with r AND rs, so I must have had some other bug that was causing this, but seems fixed now.